### PR TITLE
Fixing an edge case when using objects as constraints

### PR DESCRIPTION
**Pull request by simonc**
Hello 😊 

### Summary

This PR fixes an issue when the following situation occurs.
If you define a class like this:

```ruby
class MyConstraint
  def call(*args)
    # for some reason this is defined
  end

  def matches?(*args)
    # checking the args
  end
end
```

and try to use it as a constraint:

```ruby
get "/", to: "home#show", constraints: MyConstraint.new
```

if its `matches?` method returns `false` there will be an error because the mapper will ask for the constraint arity, thinking it is a proc, lambda or method.

The new code checks for the presence of the `arity` method on the constraint, calling it only if present, preventing the error while keeping the basic behavior.

### CODE REVIEW

**Comment by Edouard-chin** [comment]
Isn't the problem going to be the same if the call `method` of your class except a determined number of argument? (the mapper will always pass 2 arguments to the method).

Maybe it would be be better to check if the object is a proc instead of looking if it responds to `call`. This is also inline with what the documentation state (either pass a proc, either a class that responds to `matches?`) 
https://github.com/rails/rails/pull/34004/files#diff-46876897330259072a679792c5f45f99L41

File: actionpack/lib/action_dispatch/routing/mapper.rb [hunk 1]
```diff
@@ -48,10 +48,14 @@ def serve(req)
...
-          def constraint_args(constraint, request)
-            constraint.arity == 1 ? [request] : [request.path_parameters, request]
+        def constraint_args(constraint, request)
+          if constraint.respond_to?(:arity) && constraint.arity == 1
+            [request]
+          else
+            [request.path_parameters, request]
           end
+        end
+        private :constraint_args
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/routing/mapper.rb]: Why did you changed the way the private method is defined?

Comment (simonc) [reply actionpack/lib/action_dispatch/routing/mapper.rb]: @rafaelfranca I changed it because it was like that in other places of the same file 😊

Comment (simonc) [reply actionpack/lib/action_dispatch/routing/mapper.rb]: Lines 187 and 192 for instance.

**Comment by simonc** [comment]
@Edouard-chin I thought about it but didn't want to make too big of a change. I suppose I could simply ask for the arity of the `call` method instead of asking the object directly 🤔 

Something like
```ruby
def constraint_args(constraint, request)
  arity =
    if constraint.respond_to?(:arity)
      constraint.arity
    else
      constraint.method(:call).arity
    end

  return [] if arity < 1
  return [request] if arity == 1
  [request.path_parameters, request]
end
```

But I guess this doesn't cover the case where the `call` method is unrelated and expects 3 arguments but I think we're getting to the edge of an edge-case 😅

File: actionpack/lib/action_dispatch/routing/mapper.rb (hunk 1 above)
Comment (rafaelfranca) [reply actionpack/lib/action_dispatch/routing/mapper.rb]: Better to leave as it is. Line 255 is using the previous code.

**Comment by rafaelfranca** [comment]
If the `call` is unrelated it will be called and it should respond to the number of arguments passed that `constraint_args` returns, so I don't think a constraint can have a unrelated `call` and expect 3+ arguments. I think we We should check the arty of call.

**Comment by simonc** [comment]
@rafaelfranca alright, I'll update the code to the one in my previous comment. Any convention regarding the `if` alignment or can I copy/paste my comment?

**Comment by rafaelfranca** [comment]
Merged in 6a73faa5c4

**Pull request closed without merging**